{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from pathlib import Path\n",
    "\n",
    "from Bio.Align.Applications import ClustalOmegaCommandline\n",
    "import subprocess\n",
    "from Bio.SubsMat import MatrixInfo\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Seq import Seq\n",
    "from Bio import AlignIO\n",
    "\n",
    "from alignment.sequence import Sequence\n",
    "from alignment.vocabulary import Vocabulary\n",
    "from alignment.sequencealigner import SimpleScoring, LocalSequenceAligner\n",
    "\n",
    "from Bio import AlignIO\n",
    "import tables\n",
    "from Bio import Align\n",
    "\n",
    "from io import StringIO\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# # Translating and aligning GISAID sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script does a couple of things. Let's go step by step: \n",
    "    1. I take the GISAID DNA FASTA fil, translate it to protein \n",
    "        given the open reading frames. \n",
    "    2. The results are written to a file in FASTA format. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After this script, I use EMBOSS needleall tool to align one canonical \n",
    "Uniprot protein against the translated ORFs. The reuslt is an MSF file. \n",
    "\n",
    "By canonical I mean the proteins that are available on the Uniprot database: \n",
    "    https://covid-19.uniprot.org/uniprotkb?query=*\n",
    "    \n",
    "    \n",
    "The EMBOSS Needle tool that i use here specifically is the needleall tool, \n",
    "source: \n",
    "    http://emboss.sourceforge.net/apps/release/6.6/emboss/apps/needle.html\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's get to work now. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_dna(genome_records, out_file_path):\n",
    "\n",
    "    out_file_path.mkdir(parents=True, \n",
    "                        exist_ok=True)\n",
    "\n",
    "    records = []\n",
    "\n",
    "    print(\"Initiating translation...\")\n",
    "    \n",
    "    for genome_record, num in zip(genome_records, \n",
    "                                range(len(genome_records) + 1)): \n",
    "                                \n",
    "        genome_record.seq = Seq(re.sub(r'(\\W)', '', str(genome_record.seq)))\n",
    "                \n",
    "        saved_records = []\n",
    "\n",
    "        orf_proteins = find_orfs_with_trans(genome_record.seq)\n",
    "\n",
    "        for protein, i in zip(orf_proteins, range(len(orf_proteins) + 1)):\n",
    "            translated_pt_record = SeqRecord(Seq(protein), id=f\"{num}gen_{i}_orf_{genome_record.id}\", \n",
    "                                            description=f\"{num}gen_{i}_orf_{genome_record.description}\",\n",
    "                                            name=f\"{num}gen_{i}_orf_{genome_record.name}\")\n",
    "\n",
    "            saved_records.append(translated_pt_record)\n",
    "\n",
    "        if len(saved_records) > 2: \n",
    "            records.append(saved_records)\n",
    "    print(\"Genome record translated.\")\n",
    "    return records\n",
    "\n",
    "\n",
    "\n",
    "def pad_seq(sequence):\n",
    "    \"\"\" Pad sequence to multiple of 3 with N \"\"\"\n",
    "\n",
    "    remainder = len(sequence) % 3\n",
    "\n",
    "    return sequence if remainder == 0 else sequence + Seq('N' * (3 - remainder))\n",
    "\n",
    "\n",
    "def find_orfs_with_trans(seq, trans_table=1, min_protein_length=100):\n",
    "    \"\"\"Code from the Biopython library for finding open reading frames \n",
    "\n",
    "    Args: \n",
    "        seq [str]: Protein sequence\n",
    "        trans_table [int]: Look-up table \n",
    "        min_protein_length [int]: Minimum protein length\n",
    "\n",
    "    Returns: \n",
    "        answer [list]: List of protein sequences with different reading frames\n",
    "\n",
    "    \"\"\"\n",
    "    answer = []\n",
    "    seq = pad_seq(seq)\n",
    "    seq_len = len(seq)\n",
    "\n",
    "    for strand, nuc in [(+1, seq), (-1, seq.reverse_complement())]:\n",
    "        for frame in range(3):\n",
    "            trans = str(nuc[frame:].translate(trans_table))\n",
    "            trans_len = len(trans)\n",
    "            aa_start = 0\n",
    "            aa_end = 0\n",
    "            while aa_start < trans_len:\n",
    "                aa_end = trans.find(\"*\", aa_start)\n",
    "                if aa_end == -1:\n",
    "                    aa_end = trans_len\n",
    "                if aa_end - aa_start >= min_protein_length:\n",
    "                    if strand == 1:\n",
    "                        start = frame + aa_start * 3\n",
    "                        end = min(seq_len, frame + aa_end * 3 + 3)\n",
    "                    else:\n",
    "                        start = seq_len - frame - aa_end * 3 - 3\n",
    "                        end = seq_len - frame - aa_start * 3\n",
    "                    answer.append(trans[aa_start:aa_end])\n",
    "                aa_start = aa_end + 1\n",
    "    return answer\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initiating translation...\n",
      "Genome record translated.\n"
     ]
    }
   ],
   "source": [
    "genome_path = \"/home/nadzhou/SEQs/CoV/raw_seqs/gisaid10.fasta\"\n",
    "uniprot_record = SeqIO.read(\"spike_uniprot.fasta\", \"fasta\")\n",
    "genome_record = list(SeqIO.parse(genome_path, \"fasta\"))\n",
    "\n",
    "gisaid_results_path = Path(\"gisaid_results/\")\n",
    "\n",
    "results = extract_dna(genome_record, gisaid_results_path)\n",
    "\n",
    "with open(gisaid_results_path/\"translated.fasta\", \"w\") as file: \n",
    "    for record in results: \n",
    "        SeqIO.write(record, file, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The tranlsated sequence can be viewed here in this notebooks file. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EMBOSS Needleall part\n",
    "\n",
    "Now let's globally align the canonical Uniprot sequence with the \n",
    "translated sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Emboss.Applications import NeedleallCommandline\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def emboss_needle(seq_a, seq_b, out_file): \n",
    "    \"\"\"Do a global pairwise alignment using EMBOSS\n",
    "\n",
    "    Args: \n",
    "        seq_a [str]: First sequence \n",
    "        seq_b [str]: second sequence\n",
    "        out_file [str]: Output file \n",
    "\n",
    "    Returns: \n",
    "        r [subprocess object]: Execute the commandline command for EMBOSS\n",
    "    \n",
    "    \"\"\"\n",
    "    needle_cline = NeedleallCommandline(asequence=seq_a,\n",
    "                                        bsequence=seq_b, \n",
    "                                        outfile=out_file, \n",
    "                                        verbose=True,\n",
    "                                        gapextend=1,    \n",
    "                                        gapopen=10)\n",
    "\n",
    "\n",
    "    cmd = str(needle_cline)\n",
    "    cmd = cmd.split(\" \")\n",
    "    cmd.append(\"-aformat=msf\")\n",
    "    print(cmd)\n",
    "\n",
    "    r = subprocess.Popen(cmd)\n",
    "    if r.communicate(): \n",
    "        print(\"Global alignment done.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['needleall', '-verbose', '-outfile=gisaid_results/needle.fasta', '-asequence=spike_uniprot.fasta', '-bsequence=gisaid_results/translated.fasta', '-gapopen=10', '-gapextend=1', '-aformat=msf']\n",
      "Global alignment done.\n"
     ]
    }
   ],
   "source": [
    "seq_a = \"spike_uniprot.fasta\"\n",
    "seq_b = \"gisaid_results/translated.fasta\"\n",
    "out_file = \"gisaid_results/needle.fasta\"\n",
    "\n",
    "emboss_needle(seq_a, seq_b, out_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "def _identity_calc(array):\n",
    "    \"\"\"Calculate idenity if both sequences have a match my_count it up. \n",
    "    \n",
    "    Args: \n",
    "        array [nd array]: Pair of protein sequence characters\n",
    "\n",
    "    Returns: \n",
    "        my_count [int]: Hit value, 1 for identical, 0 for else\n",
    "        trimmed_seq1 [char]: Character from the first sequence\n",
    "        seq2 [char]: Character formt he second sequence\n",
    "\n",
    "    \"\"\"\n",
    "    my_count = 0\n",
    "    seq2 = \"\"\n",
    "\n",
    "    if array[0] == array[1] and array[0] != \"-\":\n",
    "        seq2 = array[1]\n",
    "        my_count += 1\n",
    "\n",
    "    return my_count, seq2\n",
    "\n",
    "\n",
    "\n",
    "def percent_id_calc(whole_seq1, whole_seq2, seq_tag, canonical_len):\n",
    "    \"\"\"Calculate percent identity of given two sequences\n",
    "    \n",
    "    Args: \n",
    "        whole_seq1 [seqrecord object]: First protein sequence\n",
    "        whole_seq2 [seqrecord]: Second protein sequence\n",
    "        seq_tag [int]: Tag for the protein sequence used \n",
    "                        for writing the file name\n",
    "    \n",
    "    \"\"\"\n",
    "    np_seq = np.asarray((whole_seq1.seq, whole_seq2.seq))\n",
    "\n",
    "    my_count, trimmed_seq2 = np.apply_along_axis(_identity_calc, 0, np_seq)\n",
    "    my_count = np.asarray(my_count, int)\n",
    "\n",
    "\n",
    "    if not np.all(my_count==0): \n",
    "        trimmed_seq2 = \"\".join(item for item in trimmed_seq2.astype(str))\n",
    "        identity_score = np.true_divide(sum(my_count), canonical_len)\n",
    "\n",
    "\n",
    "        return write_trimmed_seqs(my_count, trimmed_seq2, whole_seq1, \n",
    "                                whole_seq2, identity_score, seq_tag)\n",
    "\n",
    "    \n",
    "\n",
    "def write_trimmed_seqs(my_count, trimmed_seq2, whole_seq1,\n",
    "                        whole_seq2, identity_score, seq_tag): \n",
    "\n",
    "    \"\"\"Write the sequences that have identity scores greater than 80 percent. \n",
    "    \n",
    "    Args: \n",
    "        my_count [int]: Sum of identity scores \n",
    "        trimmed_seq1 [str]: First trimmed sequence\n",
    "        trimmed_seq2 [str]: Second trimmed sequence\n",
    "        whole_trimmed_seq1 [str]: First wholeprotein sequence\n",
    "        whole_seq2 [str]: Second protein sequence\n",
    "\n",
    "    \"\"\"\n",
    "    if identity_score > 0.7 and len(trimmed_seq2) > 50:\n",
    "        print(f\"hit hit {identity_score}\")\n",
    "\n",
    "        seq_record2 = SeqRecord(Seq(trimmed_seq2), \n",
    "                                    id=whole_seq2.id, \n",
    "                                    name=whole_seq2.name,\n",
    "                                    description=whole_seq2.description)\n",
    "\n",
    "        return seq_record2\n",
    "\n",
    "\n",
    "def write_trimmed_seqs(my_count, trimmed_seq2, whole_seq1,\n",
    "                        whole_seq2, identity_score, seq_tag): \n",
    "\n",
    "    \"\"\"Write the sequences that have identity scores greater than 80 percent. \n",
    "    \n",
    "    Args: \n",
    "        my_count [int]: Sum of identity scores \n",
    "        trimmed_seq1 [str]: First trimmed sequence\n",
    "        trimmed_seq2 [str]: Second trimmed sequence\n",
    "        whole_trimmed_seq1 [str]: First wholeprotein sequence\n",
    "        whole_seq2 [str]: Second protein sequence\n",
    "\n",
    "    \"\"\"\n",
    "    if identity_score > 0.7 and len(trimmed_seq2) > 50:\n",
    "        print(f\"hit hit {identity_score}\")\n",
    "\n",
    "        seq_record2 = SeqRecord(Seq(trimmed_seq2), \n",
    "                                    id=whole_seq2.id, \n",
    "                                    name=whole_seq2.name,\n",
    "                                    description=whole_seq2.description)\n",
    "\n",
    "        return seq_record2\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Taking the \"needle.fasta\" aligned file and look for regions of maximum \n",
    "identity. If the sequence is 90% identical trim the sequence and save \n",
    "it as FASTA. \n",
    "\n",
    "The sequences will then be run through MAFFT for a final alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intializing trimming of the aligned sequences...\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9806651198762568\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9474091260634184\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9273008507347255\n",
      "hit hit 0.9814385150812065\n",
      "hit hit 0.9489559164733179\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.97138437741686\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9690641918020109\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.8202406227883935\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9273008507347255\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9213907785336357\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.8327024981074943\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9806651198762568\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.962877030162413\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.979891724671307\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9280742459396751\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9265274555297757\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.925754060324826\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9273008507347255\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9675174013921114\n",
      "hit hit 0.9636504253673627\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9767981438515081\n",
      "hit hit 0.9767981438515081\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9690641918020109\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9814385150812065\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.878030303030303\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9265274555297757\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9806651198762568\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9752513534416086\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hit hit 0.91415313225058\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9814385150812065\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9775715390564579\n",
      "hit hit 0.9767981438515081\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.8352668213457076\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9822119102861562\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9845320959010054\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.982985305491106\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "hit hit 0.9837587006960556\n",
      "Trimmed sequences written to file.\n"
     ]
    }
   ],
   "source": [
    "record = list(AlignIO.parse(\"gisaid_results/needle.fasta\", \"msf\"))\n",
    "result_record = []\n",
    "\n",
    "\n",
    "print(\"Intializing trimming of the aligned sequences...\")\n",
    "\n",
    "for rec, num in zip(record[1:], range(len(record[1:]))): \n",
    "    canonical_ = rec[0]\n",
    "    canonical_len = len(canonical_.seq)\n",
    "\n",
    "    result = percent_id_calc(canonical_, rec[1], f\"{rec[0].id[:7]}_{num}\", canonical_len)\n",
    "\n",
    "    if result: \n",
    "        result_record.append(result)\n",
    "\n",
    "\n",
    "if result_record: \n",
    "    with open(\"gisaid_results/trimmed_seqs.fasta\", \"w\") as file: \n",
    "        for rec in result_record:\n",
    "            SeqIO.write(rec, file, \"fasta\")\n",
    "    print(\"Trimmed sequences written to file.\")\n",
    "\n",
    "            \n",
    "            \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Run a MAFFT alignment to pad the sequences into equal length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align.Applications import MafftCommandline\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mafft(in_file): \n",
    "\n",
    "    mafft_cline = MafftCommandline(input=in_file)\n",
    "                                                     \n",
    "    stdout, stderr = mafft_cline()\n",
    "    print(mafft_cline)\n",
    "\n",
    "    with open(\"gisaid_results/aligned.fasta\", \"w\") as file: \n",
    "        file.write(stdout)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mafft(\"gisaid_results/trimmed_seqs.fasta\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.7.4 64-bit ('base': conda)",
   "language": "python",
   "name": "python37464bitbaseconda43fadde7e58c4548aa12aceadbad2f74"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}